rs/execution_environment/src/scheduler.rs

```diff
@@ -2376,3 +2390,49 @@ fn scheduled_heap_delta_limit(
         .saturating_sub(remaining_heap_delta_reserve)
         .into()
 }
+
+/// Aggregator and observer of per-canister ingress queue latencies.
+struct CanisterIngressQueueLatencies {
+    /// Per canister observed ingress message latency sum and count.
+    latencies: BTreeMap<PrincipalId, (f64, usize)>,
+    /// Current block time.
+    time: Time,
+    /// Histogram to observe the latencies.
+    histogram: Histogram,
+}
+
+impl CanisterIngressQueueLatencies {
+    fn new(time: Time, histogram: Histogram) -> Self {
+        Self {
+            latencies: BTreeMap::new(),
+            time,
+            histogram,
+        }
+    }
+
+    /// Records the ingress queue latency of a message iff it is transitioning from
+    /// `Received` to some other state (i.e. when popped from the ingress queue).
+    fn on_ingress_status_changed(&mut self, old_status: Arc<IngressStatus>) {
```

Comment:
I guess there isn't a "smart" way to make sure, but did we cover all the relevant places where this changes (I mean where it transitions from `Received` to something else)? I guess it technically doesn't say it covers all the places but that is the intention right?

Reply:
I figured something like that, thanks for clarifying.